src/loan.rs

```diff
@@ -60,21 +60,177 @@ static COVENANT_PK: &str = "03b9b6059008e3576aad58e05a3a3e37133b05f68cda8535ec09
 /// `covenant_pk`, which is only used to verify that the transaction
 /// data on the witness stack matches the transaction which triggered
 /// the call.
-fn collateral_descriptor(
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+struct CollateralContract {
+    descriptor: Descriptor<PublicKey>,
     borrower_pk: PublicKey,
     lender_pk: PublicKey,
-    timelock: u64,
-    repayment_output: TxOut,
-) -> Result<Descriptor<PublicKey>> {
-    let repayment_output = serialize(&repayment_output).to_hex();
-    let desc = Descriptor::<elements::bitcoin::PublicKey>::from_str(
-        &(format!(
-            "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
-            COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
-        )),
-    )?;
-
-    Ok(desc)
+    timelock: u32, // TODO: Model timelocks as u32
+}
+
+impl CollateralContract {
+    /// Fill in the collateral contract template with the provided arguments.
+    fn new(
+        borrower_pk: PublicKey,
+        lender_pk: PublicKey,
+        timelock: u64,
+        repayment_output: TxOut,
+    ) -> Result<Self> {
+        let repayment_output = serialize(&repayment_output).to_hex();
+        let descriptor = Descriptor::<elements::bitcoin::PublicKey>::from_str(
+            &(format!(
+                "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
+                COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
+            )),
+        )
+        .context("invalid collateral output descriptor")?;
+
+        Ok(Self {
+            descriptor,
+            borrower_pk,
+            lender_pk,
+            timelock: timelock as u32,
+        })
+    }
+
+    fn address(&self) -> Address {
+        // FIXME: AddressParams should not be hard-coded
+        self.descriptor
+            .address(&AddressParams::ELEMENTS)
+            .expect("descriptor address to exist")
+    }
+
+    fn blinded_address(&self, blinder: secp256k1_zkp::PublicKey) -> Address {
+        self.descriptor
+            .blind_addr(Some(blinder), &AddressParams::ELEMENTS)
+            .expect("descriptor address to exist")
+    }
+
+    async fn satisfy_loan_repayment<S, SF>(
+        &self,
+        transaction: &mut Transaction,
+        input_value: confidential::Value,
+        input_index: u32,
+        identity_signer: S,
+    ) -> Result<()>
+    where
+        S: FnOnce(secp256k1::Message) -> SF,
+        SF: Future<Output = Result<Signature>>,
+    {
+        let descriptor_cov = &self.descriptor.as_cov().expect("covenant descriptor");
+
+        let cov_script = descriptor_cov.cov_script_code();
+        let transaction_cloned = transaction.clone();
```

Comment:
Good question! The problem is that we use a mutable reference to the transaction, so we cannot also use an immutable one at the same time. We therefore clone the transaction so that we can have an immutable reference to the cloned value and still mutate the original.

There might be a better solution, but I couldn't find it.

Reply:
Oh, I did not see that it was a mutable reference.